torch_xla_py/xla_dist.py

```diff
@@ -402,3 +407,241 @@ def get_cluster(self):
+              worker_idx=idx, worker_ip=service_worker._internal_ip,
+              worker_port=service_worker._port )
+          for idx, service_worker in enumerate(self._cluster._service_workers)
+      ]
+      xrt_tpu_config = '|'.join(xrt_server_config)
+      env_vars['XRT_TPU_CONFIG'] = '"{}"'.format(xrt_tpu_config)
+
+    export_cmd = ''
+    for k in env_vars:
+      export_cmd += 'export {}={}; '.format(k, env_vars[k])
+
+    return ''.join([export_cmd, cmd])
+
+  def _prepare_scripts(self, cmd):
+    for var in self.DIST_ENV_VARS:
+      if var in cmd:
+        raise ValueError((
+            '{} should not be in the training command provided as they'
+            ' will interfere with the values set for distributed'
+            ' training'.format(var)))
+
+    worker_script_map = {}
+    for i in range(len(self._cluster._client_workers)):
+      script_path = self.SCRIPT_PATH_TMPL.format(worker=i)
+      script = self._export_env_vars(cmd, i)
+      if self.is_docker:
+        script = self._add_docker_vars(script)
+      # Setup environment for non-interactive non-login shell over ssh
+      script = '{}; {}'.format('. /etc/profile', script)
+
+      with open(script_path, 'w') as f:
+        f.write(script)
+      subprocess.call(['chmod', '+x', script_path])
+      worker_script_map[self._cluster._client_workers[i]] = script_path
+
+    return worker_script_map
+
+  def _scp_scripts(self, script_map):
+
+    def _gcloud_scp(script_path, client_worker):
+      scp_cmd = [
+          'gcloud', '-q', 'compute', 'scp', '--internal-ip',
+          '--zone={}'.format(client_worker._zone),
+          script_path, 'pytorchtpudistrunner@{}:~/{}'.format(
+              client_worker._hostname, os.path.basename(script_path)),
+      ]
+      proc = subprocess.Popen(scp_cmd, stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE)
+      self._stream_logs(proc, client_worker)
+
+    threads = []
+    for i, client_worker in enumerate(script_map):
+      if i == 0:
+        # ssh keygen single time
+        _gcloud_scp(script_map[client_worker], client_worker)
+        continue
+      thread = threading.Thread(
+          target=_gcloud_scp,
+          args=(script_map[client_worker], client_worker,))
+      thread.daemon = True
+      thread.start()
+      threads.append(thread)
+
+    for thread in threads:
+      thread.join()
+
+  def _rm_docker_container(self, client_worker):
+    rm_cmd = ('gcloud -q compute ssh --internal-ip --zone={}'
+              ' pytorchtpudistrunner@{} --command="docker rm -f {}"').format(
+                  client_worker._zone, client_worker._hostname,
+                  self.container_name)
+    proc = subprocess.Popen(rm_cmd, stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE, shell=True)
+    self._stream_logs(proc, client_worker)
+
+  def _start_run(self, script_map):
+
+    def _gcloud_ssh(script_path, client_worker, event):
+      while event.is_set():
```

Comment:
This won't work.
You will be sitting in the stream-log thread join, and this will never be noticed if changes.

Reply:
AFAICT each thread will enter `_gcloud_ssh()`, check that the event is set (it should be at entry), then create an SSH process (Popen) and sit waiting for its logs (`stdout.join()`).
When you clear the event in the Ctrl^C handler, the `_gcloud_ssh()` thread won't notice since it already past the event check.